### PR TITLE
FreeDV 2020 support

### PR DESCRIPTION
**Pull request by ac0zj**
implementing FreeDV 2020 mode into codec2 repo, mainly freedv_api.c changes, and linking in LPCNet.

### CODE REVIEW

File: src/freedv_api.h [hunk 1]
```diff
@@ -49,6 +49,7 @@
 #define FREEDV_MODE_800XA       5
 #define FREEDV_MODE_700C        6
 #define FREEDV_MODE_700D        7
+#define FREEDV_MODE_2020        8
```
Comment (db4ple) [inline src/freedv_api.h]: Would be nice if you also would add the statement 
```
#if !defined(FREEDV_MODE_2020_EN)
        #define FREEDV_MODE_2020_EN FREEDV_MODE_EN_DEFAULT
#endif
```
See a couple of linces below.

Comment (ac0zj) [reply src/freedv_api.h]: Thanks, I added that to brad-2020.

**Comment by drowe67** [comment]
Note to all - I'm still buidling up LPCNet support, need to:
1.  Get Travis to build the codec2/LPCNet repo and link to it.  Looks like Travis is OK with the default (LPCNet #ifdef-ed out).
2. Fill out the LPCNet API functions.
3. Complete the freedv_api functions

File: src/freedv_api.h [hunk 2]
```diff
@@ -87,8 +91,9 @@
 #if !defined(FREEDV_MODE_800XA_EN)
         #define FREEDV_MODE_800XA_EN FREEDV_MODE_EN_DEFAULT
 #endif
-
-
+#if !defined(FREEDV_MODE_2020_EN)
+        #define FREEDV_MODE_2020_EN FREEDV_MODE_EN_DEFAULT
+#endif
```
Comment (drowe67) [inline src/freedv_api.h]: oops like look I doubled up on this!

**Comment by drowe67** [comment]
@hobbes1069 Hi Richard, is it possible pls to add some CMake code to set up the include and static library path for compiling and linking with LPCNet? 

This put ```liblpcnetfreedv.a``` in the wrong place on the linker line:
```
CFLAGS="-D__LPCNET__ -I /home/david/LPCNet/src" LDFLAGS="-L/home/david/LPCNet/build_linux/src/ -llpcnetfreedv" cmake ..
```
I've managed to do it manually on the command line with:
```
/usr/bin/cc -D__LPCNET__ -I /home/david/LPCNet/src -Wall -Wno-strict-overflow -std=gnu11 -fPIC -g -O2  -rdynamic CMakeFiles/freedv_tx.dir/freedv_tx.c.o  -o freedv_tx -Wl,-rpath,/home/david/codec2/build_linux/src -lm libcodec2.so.0.8 /home/david/LPCNet/build_linux/src/liblpcnetfreedv.a -lm
```

Thanks

**Comment by drowe67** [comment]
@hobbes1069 I've worked out a temporary way to link liblpcnetfreedv, so I can continue to develop FreeDV 2020 support.  Pls feel free to improve.

File: README.md [hunk 3]
```diff
@@ -84,6 +84,26 @@ $ ./src/c2enc 1300 ../raw/hts1a.raw - | ./src/c2dec 1300 - - | play -t raw -r 80
...
+   ```
+
+2. (re)build Codec 2 with LPCNet support:
+   ```
+   $ cd ~/codec2
+   $ rm -Rf build_linux/ && mkdir build_linux && cd build_linux
+   $ CFLAGS="-D__LPCNET__ -I/path/to/LPCNet/src" LDFLAGS=-L/path/to/LPCNet/build_linux/src cmake ..
+   $ make
+   ```
+
```
Comment (drowe67) [inline README.md]: @hobbes1069 Here's how I'm building everything atm, I've also changed the CMakeLists.txt file.  I'm sure there's a better way.

Comment (drowe67) [reply README.md]: Note it won't build without __LPCNET__ atm, as I've added ```lpcnetfreedv``` to every ```target_link_libraries```.  Hence Travis is choking.

File: README.md [hunk 4]
```diff
@@ -84,6 +84,49 @@ $ ./src/c2enc 1300 ../raw/hts1a.raw - | ./src/c2dec 1300 - - | play -t raw -r 80
...
+Listen the freedv_tx:
+```
+$ ./freedv_tx 2020 ~/LPCNet/wav/wia.wav - | aplay -f S16_LE
+```
+
+FreeDV API tx, with reference rx from above:
+```
+$ ./freedv_tx 2020 ~/LPCNet/wav/wia.wav - | ./ofdm_demod --nc 31 --verbose 1 --ldpc 2 -p 312 | ~/LPCNet/build_linux/src/lpcnet_dec -s | aplay -f S16_LE -r 16000
+```
+
```
Comment (drowe67) [inline README.md]: @ac0zj This section of ```README.md``` explains how to build and run ```freetx_tx``` with 2020

**Comment by drowe67** [comment]
@hobbes1069 I just about have this working but rpath is not being set so when I try to run say src/freedv_tx it can't find liblpcnetfreed.so

**Comment by hobbes1069** [comment]
Ok, wow, this is a long PR :)

I think the short answer is we need to do a real install e.g.: install(TARGET ...) and then add the required find_libary(...) in freedv.

**Comment by hobbes1069** [comment]
Ok, I committed some updates to freedv-gui/brad-2020 that fixed building for me. It's a quick fix though...

**Comment by drowe67** [comment]
> Ok, I committed some updates to freedv-gui/brad-2020 that fixed building for me. It's a quick fix though...

Thanks!  Is there anything similar we can do for this branch?  So freedv_tx has the correct rpath to liblpcnetfreedv.so while I'm developing this branch and LPCNet?

**Comment by hobbes1069** [comment]
Whoops, I conflated freedv_tx with freedv-gui but it's part of codec2... Too many irons in the fire. Working on codec2/brad-2020 now.

**Comment by hobbes1069** [comment]
Ok, I committed an update to LPCNet in master and brad-2020 for codec2. Works for me but I need to do some cleanup work at some point.

**Comment by drowe67** [comment]
> Ok, I committed an update to LPCNet in master and brad-2020 for codec2. Works for me but I need to do some cleanup work at some point.

Thanks Richard, it builds OK, but doesn't seem to find the library at run time.
```
$ cd ~/codec2
$ rm -Rf build_linux/ && mkdir build_linux/ && cd build_linux/
$ cmake -DLPCNET_BUILD_DIR=/home/david/LPCNet/build_linux ..
$ cd src
./freedv_tx 2020 ~/LPCNet/wav/wia.wav - | ./freedv_rx 2020 - - | aplay -f S16_LE -r 16000
./freedv_rx: error while loading shared libraries: liblpcnetfreedv.so: cannot open shared object file: No such file or directory
```
This would be useful for debugging without having to set up LD_LIBRARY_PATH.  Curiously, freedv-gui, when built with CODEC2_BUILD_DIR, does set the rpath correctly at link time to find libcodec2.so.  That's the bit that has me stumped.

also it won't build without LPCNet:
```
$ cd ~/codec2
$ rm -Rf build_linux/ && mkdir build_linux/ && cd build_linux/
$ cmake ..
<snip>
CMake Error at CMakeLists.txt:204 (find_package):
  By not providing "Findlpcnet.cmake" in CMAKE_MODULE_PATH this project has
  asked CMake to find a package configuration file provided by "lpcnet", but
  CMake did not find one.

  Could not find a package configuration file provided by "lpcnet" with any
  of the following names:

    lpcnetConfig.cmake
    lpcnet-config.cmake
```

**Comment by drowe67** [comment]
@ac0zj freedv_rx working for 2000 ... psl see updated README.md

**Comment by hobbes1069** [comment]
I haven't tried running any binaries yet but I did a standard "make install" which put the lpcnet stuff in /usr/local/lib64 (for Fedora). If it's not finding the library you may just need to run ldconfig.

**Comment by hobbes1069** [comment]
Oh yes, the "bootstrapping" problem. I'll work on that today.

**Comment by drowe67** [comment]
> I haven't tried running any binaries yet but I did a standard "make install" which put the lpcnet stuff in /usr/local/lib64 (for Fedora). If it's not finding the library you may just need to run ldconfig.

1. During development, it's common to change LPCNet; make, change codec2, make.  I'd rather not 'make install' every time I make a small change to the LPCNet source.  It should be possible for the codec2 build system to find (at run time) liblpcnetfreedv.so inside of LPCNet.  

   This is exactly what happens with freedv-gui when CODEC2_BUILD_DIR is set ; if I modify codec2, freedv-gui finds libcodec2.so at run time inside the codec2 diectory.  I don't have to 'make install' during development. Looking at the link line, the freedv-gui build system sets the rpath to libcodec2.so with a linker option.

   Can we pls  configure codec2 cmake to set the rpath when it builds, so it finds liblpcnetfreedv.so at run time without a make install && ldconfig, or setting LD_LIBRARY_PATH?  Just a small thing, but would make development easier.

2. Also we need to be able to build codec2 without LPCNet - currently that's not working (see example above).

3. What did you have in mind with bootstrapping?

**Comment by hobbes1069** [comment]
1. Ok, I kept your conditional in place so it should still be working unless I messed it up somehow. Using LPCNET_BUILD_DIR is probably the best option for testing repeated/incremental builds.

2. Ok so we need a bypass method. I'll have to think about that.

3. I was thinking about the way we bootstrap wxWidgets for WIN32 builds but I'd rather not do something that complicated if it can be avoided.

**Comment by hobbes1069** [comment]
The last commit is untested but should hopefully do the trick. Setting LPCNET_BUILD_DIR implies LPCNET=ON otherwise it should be skipped.

**Comment by drowe67** [comment]
>     1. Ok, I kept your conditional in place so it should still be working unless I messed it up somehow. Using LPCNET_BUILD_DIR is probably the best option for testing repeated/incremental builds.
> 
>     2. Ok so we need a bypass method. I'll have to think about that.
> 
>     3. I was thinking about the way we bootstrap wxWidgets for WIN32 builds but I'd rather not do something that complicated if it can be avoided.

1. Sorry I don't understand what you mean by the conditional?  I haven't ever been able to get a codec 2 executable to find lpcnetfreedv.so at run time without LD_LIBRARY_CONFIG and friends - pls try to run freedv_tx and you'll see what I mean:
```
/codec2/build_linux/src$ ./freedv_tx
./freedv_tx: error while loading shared libraries: liblpcnetfreedv.so: cannot open shared object file: No 
```
As explained above freedv-gui **can** find libcodec2.so, even without it being installed.  So I figure it must be possible for codec2 executables to find the liblpcnetfreedv.so somehow.
2. Looks good now - thanks.
3. I don't think anything needs to be done here.  Anyone building LPCNet and codec2 simply has to manually build and install codec2 first - instructions are on step 1 of the README.md for 2020.

**Comment by hobbes1069** [comment]
Ok, for that use case static libs may be easier than dealing with RPATH... To know why it can't find it I'd probably need the build log to see how it was linked. I'm going to guess that because you're adding a search path (-L method) that it's not storing the RPATH, but if you gave the linker the full path to the library it would "remember" it.

**Comment by hobbes1069** [comment]
Ok, when I try, I build codec2 without LPCNet, which works fine, but when I specify the build directory I get the following:
```
[  1%] Linking C executable drs232
/usr/bin/ld: /home/build/build/LPCNet/src/liblpcnetfreedv.so: undefined reference to `estimate_amplitudes'
/usr/bin/ld: /home/build/build/LPCNet/src/liblpcnetfreedv.so: undefined reference to `nlp'
/usr/bin/ld: /home/build/build/LPCNet/src/liblpcnetfreedv.so: undefined reference to `dft_speech'
/usr/bin/ld: /home/build/build/LPCNet/src/liblpcnetfreedv.so: undefined reference to `two_stage_pitch_refinement'
/usr/bin/ld: /home/build/build/LPCNet/src/liblpcnetfreedv.so: undefined reference to `c2const_create'
/usr/bin/ld: /home/build/build/LPCNet/src/liblpcnetfreedv.so: undefined reference to `est_voicing_mbe'
/usr/bin/ld: /home/build/build/LPCNet/src/liblpcnetfreedv.so: undefined reference to `nlp_destroy'
/usr/bin/ld: /home/build/build/LPCNet/src/liblpcnetfreedv.so: undefined reference to `nlp_create'
/usr/bin/ld: /home/build/build/LPCNet/src/liblpcnetfreedv.so: undefined reference to `kiss_fft_alloc'
/usr/bin/ld: /home/build/build/LPCNet/src/liblpcnetfreedv.so: undefined reference to `make_analysis_window'
```

**Comment by hobbes1069** [comment]
Ok, here's a BIG headsup! I just completely reworked both LPCnet and codec2 to properly use the cmake method of exporting and importing targets to get rid of all the manual finding of libraries and headers. I was able to build codec2, then lpcnetfreedv, and then codec2 again, linking to lpcnetfreedv and was able to run freedv_tx from the source directly. 

I'm sure the readmes need to be updated but please test first. You must use LPCENET_BUILD_DIR and CODEC2_BUILD_DIR and it will file the lpcnet.cmake and codec2.cmake files I'm now exporting.

Commits to arrive shortly.

**Comment by hobbes1069** [comment]
And I broke Travis CI... :|

File: CMakeLists.txt [hunk 5]
```diff
@@ -185,11 +185,60 @@ include_directories(${PROJECT_BINARY_DIR})
...
 #
 # codec2 library
 #
 add_subdirectory(src)
 
+
+if(UNIX)
+    target_link_libraries(codec2 m)
+endif(UNIX)
+
```
Comment (db4ple) [inline CMakeLists.txt]: The 3 lines above have to go to src/CMakeLists.txt behind add_library(codec2 ${CODEC2_SRCS}) then cmake works for me (and would in Travis)

Comment (hobbes1069) [reply CMakeLists.txt]: Ahh, you must be running an older version of CMake where it has to go in the same CMakeLists file... I'll fix it here in a second.

**Comment by db4ple** [comment]
Travis runs Ubuntu 16.04 so do I on one of my machines. It uses cmake 3.12.4

**Comment by hobbes1069** [comment]
> Travis runs Ubuntu 16.04 so do I on one of my machines. It uses cmake 3.12.4

I think this behavior was changed in 3.13 and Fedora is on 3.14.x.

**Comment by drowe67** [comment]
@hobbes1069 it's not working for me (cmake 3.9.1), but could be user error. Can you pls document the steps required?  This is what I'm doing:

```
cd ~/codec2
rm -Rf build_linux/ && mkdir build_linux/ && cd build_linux/
cmake .. && make
cd ~/LPCNet
rm -Rf build_linux/ && mkdir build_linux/ && cd build_linux/
cmake -DCODEC2_BUILD_DIR=~/codec2/build_linux .. && make
cd ~/codec2
rm -Rf build_linux/ && mkdir build_linux/ && cd build_linux/
cmake -DLPCNET_BUILD_DIR=~/LPCNet/build_linux ..
<snip>
CMake Error at CMakeLists.txt:221 (target_link_libraries):
  Attempt to add link library "lpcnetfreedv" to target "codec2" which is not
  built in this directory.
<snip>
-- Configuring incomplete, errors occurred!
```
Let me know if you want the log files.

Thanks for working on this!

**Comment by db4ple** [comment]
Hi, @drowe67 , @hobbes1069 : I am pretty sure, it is the same issue we had with the previous breakdown (the wrong location of target_link_libraries in case of older cmake versions)

**Comment by hobbes1069** [comment]
Not that it should make a difference in this case, but you don't have to delete the codec2 build tree. When you run cmake again with new parameters it knows what needs to be rebuilt.

He's right, it doesn't like the directive being in the main CMakeLists.txt but the add_library is in src/CMakeLists.txt... I'll move it over but it was nice and pretty all the logic being in one place :)

**Comment by drowe67** [comment]
@hobbes1069 

Hi Richard when I try:
```
cmake -DLPCNET_BUILD_DIR=~/LPCNet/build_linux ..
grep LPCNet CMakeCache.txt
CMakeCache.txt://Build codec2 with LPCNet support.
CMakeCache.txt:LPCNET:BOOL=OFF
CMakeCache.txt:LPCNET_BUILD_DIR:PATH=/home/david/LPCNet/build_linux
CMakeCache.txt:lpcnet_DIR:PATH=/home/david/LPCNet/build_linux
```

When I  ```make``` I can't see ```__LPCNET__```  defined, and freedv_tx/rx do not support 2020 (I get an assert).

Can you pls:
1. Confirm that you have built codec2 with LPCNet and that freedv_tx/freetx_rx with the 2020 options works, for example:
   ```
   cd codec2/build_linux/src
   ./freedv_tx 2020 ~/LPCNet/wav/all.wav - | ./freedv_rx 2020 - - | aplay -f S16_LE -r 16000
   ```
2. Provide specific instructions on how we can build from a fresh machine in README.md

**Comment by hobbes1069** [comment]
I just verified it still works for me... add_definitions(...) hasn't changed as far as I know. I guess I need to take you up on the Mint login.

**Comment by drowe67** [comment]
@donmr @db4ple I'd like to merge this branch in the next few days, as it contains some 700D bug fixes as well as 2020 support. I've just merged it with master so it should be up to date. Can you pls tell me if it works OK with yr stm32 tests?

**Comment by db4ple** [comment]
Docker says yes on bear 👍

**Comment by drowe67** [comment]
> Docker says yes on bear +1

Thanks @db4ple. @donmr when convenient, pls let me know if you are OK with merging this from a stm32 point of view.

**Comment by donmr** [comment]
Compiles and runs my tests fine.

**Pull request merged**